File: demo-ios-objc/demo-ios-objc/IdentifyViewController.m
//
//  IdentifyViewController.m
//  demo-ios-objc
//
//  Copyright © 2020 Optable Technologies Inc. All rights reserved.
//  See LICENSE for details.
//

#import "OptableSDKDelegate.h"
#import "IdentifyViewController.h"
#import "AppDelegate.h"
#import "demo_ios_objc-Swift.h"

@import OptableSDK;

@interface IdentifyViewController ()
@property (nonatomic, strong, nullable) id networkLogObserver;
@end

@implementation IdentifyViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _identifyInput.delegate = self;
    
    OptableSDKDelegate *delegate = (OptableSDKDelegate *)OPTABLE.delegate;
    delegate.identifyOutput = self.identifyOutput;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear: animated];
    [self startObservingNetworkLogs];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear: animated];
    [self stopObservingNetworkLogs];
}

// MARK: - Actions
- (IBAction)dispatchIdentify:(id)sender {
    [self.view endEditing: TRUE];
    
    NSString *email = _identifyInput.text;
    
    NSMutableString *output = [NSMutableString stringWithFormat: @"Calling /identify API with:\n\n"];
    if (email.length > 0) {
        [output appendString: [NSString stringWithFormat: @"Email: %@\n", email]];
    }
    
    _identifyOutput.text = output;
    
    NSError *error = nil;
    
    [OPTABLE identify: @[
        [OptableSDKIdentifier identifierWithType:OptableSDKIdentifierType_EmailAddress value:email],
        [OptableSDKIdentifier identifierWithType:OptableSDKIdentifierType_PhoneNumber value:@"+1234567890"],
        [OptableSDKIdentifier identifierWithType:OptableSDKIdentifierType_AppleIDFA value:@"06DE8C6A-A431-4235-A262-E3A9C2CCEB34"],
        [OptableSDKIdentifier identifierWithType:OptableSDKIdentifierType_GoogleGAID value:@"D04BB8C3-5A3E-4964-9757-D38365F59E6A"],
        [OptableSDKIdentifier identifierWithRawType:@"c" value:@"new-custom.ABC"],
        [OptableSDKIdentifier identifierWithType:OptableSDKIdentifierType_Custom value:@"custom-7" customIdx:@7],
        [OptableSDKIdentifier identifierWithString:@"c9:custom-9-id"],
    ]
                error: &error];
}

// MARK: - UITextFieldDelegate
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return FALSE;
}

// MARK: - Logging
- (void)startObservingNetworkLogs {
    __weak typeof(self) weakSelf = self;
    
    _networkLogObserver = [NSNotificationCenter.defaultCenter
                           addObserverForName: NotificationNames.HTTPURLLogUpdated
                           object: nil
                           queue: [NSOperationQueue mainQueue]
                           usingBlock: ^(NSNotification* notification) {
        HTTPURLLogEntry *logEntry = notification.userInfo[@"data"];
        if (![logEntry isKindOfClass:[HTTPURLLogEntry class]]) {
            return;
        }
        
        NSString *urlString = logEntry.request.URL.absoluteString;
        if ([urlString containsString:@"/identify"]) {
            weakSelf.identifyOutput.text = logEntry.debugDescription;
            
            if (logEntry.response == nil) {
                NSLog(@"%@", logEntry.requestDebugDescription);
            } else {
                NSLog(@"%@", logEntry.responseDebugDescription);
            }
        }
    }];
}

- (void)stopObservingNetworkLogs {
    if (_networkLogObserver != NULL) {
        [NSNotificationCenter.defaultCenter removeObserver: _networkLogObserver];
    }
}

@end

File: demo-ios-swift/demo-ios-swift/IdentifyViewController.swift
//
//  IdentifyViewController.swift
//  demo-ios-swift
//
//  Copyright © 2020 Optable Technologies Inc. All rights reserved.
//  See LICENSE for details.
//

import OptableSDK
import UIKit
import AppTrackingTransparency

// MARK: - IdentifyViewController
class IdentifyViewController: UIViewController {
    // Outlets
    @IBOutlet var identifyInput: UITextField!
    @IBOutlet var identifyButton: UIButton!
    @IBOutlet var identifyOutput: UITextView!

    // Logging
    private var networkLogObserver: (any NSObjectProtocol)?

    override func viewDidLoad() {
        super.viewDidLoad()
        identifyInput.delegate = self
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        startObservingNetworkLogs()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        stopObservingNetworkLogs()
    }

    // MARK: Actions
    // dispatchIdentify() is the action invoked on a click on the "Identify" UIButton in our demo app.
    // It initiates a call to the OptableSDK.identify() API and prints debugging information to the UI and debug console.
    @IBAction func dispatchIdentify(_ sender: UIButton) {
        view.endEditing(true)
        
        // If еру user gives consent to be tracked - IDFA will be added
        // automatically to the list of identifiers,
        // either in the `identify` or `targeting` calls
        ATTrackingManager.requestTrackingAuthorization { [self] _ in
            DispatchQueue.main.async {
                do {
                    let email = self.identifyInput.text ?? ""
                    let optableIdentifiers: [OptableIdentifier] = [
                        .emailAddress(email),
                        .phoneNumber("+1234567890"),
                        .custom(nil, "new-custom.ABC"),
                        .custom(9, "custom-9-id")
                    ]
                    
                    try OPTABLE!.identify(optableIdentifiers) { result in
                        switch result {
                        case .success:
                            print("[OptableSDK] ✅ Success on /identify API call")
                        case let .failure(error):
                            print("[OptableSDK] 🚫 Error on /identify API call: \(error)")
                        }
                    }
                    
                } catch {
                    print("[OptableSDK] 🚫 Exception on /identify API call: \(error)")
                    self.identifyOutput.text += "🚫 EXCEPTION: \(error)"
                }
            }
        }
        
    }
}

// MARK: - UITextFieldDelegate
extension IdentifyViewController: UITextFieldDelegate {
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return false
    }
}

// MARK: - Logging
private extension IdentifyViewController {
    /// Setups observation for URLRequest/URLResponse logging
    func startObservingNetworkLogs() {
        networkLogObserver = NotificationCenter.default
            .addObserver(forName: .HTTPURLLogUpdated, object: nil, queue: .main, using: { [weak self] notification in
                if let logEntry = notification.userInfo?["data"] as? HTTPURLLogEntry,
                   logEntry.request.url?.absoluteString.contains("/identify") == true {
                    self?.identifyOutput.text = logEntry.debugDescription
                    print(logEntry.response == nil ? logEntry.requestDebugDescription : logEntry.responseDebugDescription)
                }
            })
    }

    func stopObservingNetworkLogs() {
        guard let networkLogObserver else { return }
        NotificationCenter.default.removeObserver(networkLogObserver)
    }
}
